OpenRA.Mods.Common/Activities/FindAndDeliverResources.cs

```diff
@@ -194,6 +198,21 @@ public override bool Tick(Actor self)
 					if ((loc - searchFromLoc.Value).LengthSquared > searchRadiusSquared)
 						return int.MaxValue;
 
+					// Add a cost modifier to harvestable cells to prefer resources that are closer to the refinery.
+					// This reduces the tendancy for harvesters to move in straight lines
+					if (procPos.HasValue && harvInfo.ResourceRefineryDirectionCostModifier > 0 && harv.CanHarvestCell(self, loc))
+					{
+						var pos = self.World.Map.CenterOfCell(loc);
+
+						// Calculate harv-cell-refinery angle (cosine rule)
+						var aSq = (harvPos - procPos.Value).LengthSquared;
+						var bSq = (pos - procPos.Value).LengthSquared;
+						var cSq = (pos - harvPos).LengthSquared;
+						var cosA = (int)(1024 * (bSq + cSq - aSq) / (2 * Exts.ISqrt(bSq * cSq)));
+
+						return harvInfo.ResourceRefineryDirectionCostModifier * cosA / 1024;
```

Comment:
This may run the risk of the total cost becoming negative. I don't know what the consequences of that are, but it may break things. Perhaps it is better to add a fixed number so that the cost is 0 when moving directly towards the refinery.